File: Assets/zfoolua/common/TripleLong.lua

local TripleLong = {}

function TripleLong:new()
    local obj = {
        left = 0, -- long
        middle = 0, -- long
        right = 0 -- long
    }
    setmetatable(obj, self)
    self.__index = self
    return obj
end

function TripleLong:protocolId()
    return 114
end

function TripleLong:write(buffer, packet)
    if packet == nil then
        buffer:writeInt(0)
        return
    end
    buffer:writeInt(-1)
    buffer:writeLong(packet.left)
    buffer:writeLong(packet.middle)
    buffer:writeLong(packet.right)
end

function TripleLong:read(buffer)
    local length = buffer:readInt()
    if length == 0 then
        return nil
    end
    local beforeReadIndex = buffer:getReadOffset()
    local packet = TripleLong:new()
    local result0 = buffer:readLong()
    packet.left = result0
    local result1 = buffer:readLong()
    packet.middle = result1
    local result2 = buffer:readLong()
    packet.right = result2
    if length > 0 then
        buffer:setReadOffset(beforeReadIndex + length)
    end
    return packet
end

return TripleLong

File: Assets/zfoocs/Common/TripleLong.cs
using System;
using System.Collections.Generic;
namespace zfoocs
{
    
    public class TripleLong
    {
        public long left;
        public long middle;
        public long right;
    }

    public class TripleLongRegistration : IProtocolRegistration
    {
        public short ProtocolId()
        {
            return 114;
        }
    
        public void Write(ByteBuffer buffer, object packet)
        {
            if (packet == null)
            {
                buffer.WriteInt(0);
                return;
            }
            TripleLong message = (TripleLong) packet;
            buffer.WriteInt(-1);
            buffer.WriteLong(message.left);
            buffer.WriteLong(message.middle);
            buffer.WriteLong(message.right);
        }
    
        public object Read(ByteBuffer buffer)
        {
            int length = buffer.ReadInt();
            if (length == 0)
            {
                return null;
            }
            int beforeReadIndex = buffer.GetReadOffset();
            TripleLong packet = new TripleLong();
            long result0 = buffer.ReadLong();
            packet.left = result0;
            long result1 = buffer.ReadLong();
            packet.middle = result1;
            long result2 = buffer.ReadLong();
            packet.right = result2;
            if (length > 0)
            {
                buffer.SetReadOffset(beforeReadIndex + length);
            }
            return packet;
        }
    }
}